flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/CookieHandler.java

```diff
@@ -0,0 +1,130 @@
+			super.channelActive(ctx);
+			LOG.debug("In channelActive method of ClientCookieHandler");
+
+			if(this.secureCookie != null && this.secureCookie.length() != 0) {
+				LOG.debug("In channelActive method of ClientCookieHandler -> sending secure cookie");
+				final ByteBuf buffer = Unpooled.buffer(4 + this.secureCookie.getBytes(DEFAULT_CHARSET).length);
+				buffer.writeInt(secureCookie.getBytes(DEFAULT_CHARSET).length);
+				buffer.writeBytes(secureCookie.getBytes(DEFAULT_CHARSET));
+				ctx.writeAndFlush(buffer);
+			}
+		}
+	}
+
+	public static class ServerCookieDecoder extends MessageToMessageDecoder<ByteBuf> {
+
+		private final String secureCookie;
+
+		private final List<Channel> channelList = new ArrayList<>();
+
+		private final Charset DEFAULT_CHARSET = Charset.forName("utf-8");
+
+		private final Logger LOG = LoggerFactory.getLogger(ServerCookieDecoder.class);
+
+		public ServerCookieDecoder(String secureCookie) {
+			this.secureCookie = secureCookie;
+		}
+
+		/**
+		 * Decode from one message to an other. This method will be called for each written message that can be handled
+		 * by this encoder.
+		 *
+		 * @param ctx the {@link ChannelHandlerContext} which this {@link MessageToMessageDecoder} belongs to
+		 * @param msg the message to decode to an other one
+		 * @param out the {@link List} to which decoded messages should be added
+		 * @throws Exception is thrown if an error accour
+		 */
+		@Override
+		protected void decode(ChannelHandlerContext ctx, ByteBuf msg, List<Object> out) throws Exception {
+
+			LOG.debug("ChannelHandlerContext name: {}, channel: {}", ctx.name(), ctx.channel());
+
+			if(secureCookie == null || secureCookie.length() == 0) {
+				LOG.debug("Not validating secure cookie since the server configuration is not enabled to use cookie");
+				return;
+			}
+
+			LOG.debug("Going to decode the secure cookie passed by the remote client");
+
+			if(channelList.contains(ctx.channel())) {
+				LOG.debug("Channel: {} already authorized", ctx.channel());
+				return;
+			}
+
+			//read cookie based on the cookie length passed
+			int cookieLength = msg.readInt();
+			if(cookieLength != secureCookie.getBytes(DEFAULT_CHARSET).length) {
+				ctx.channel().close();
+				String message = "Cookie length does not match with source cookie. Invalid secure cookie passed.";
+				throw new IllegalStateException(message);
+			}
+
+			//read only if cookie length is greater than zero
+			if(cookieLength > 0) {
+
+				final byte[] buffer = new byte[secureCookie.getBytes(DEFAULT_CHARSET).length];
+				msg.readBytes(buffer, 0, cookieLength);
+
+				if(!Arrays.equals(secureCookie.getBytes(DEFAULT_CHARSET), buffer)) {
+					LOG.error("Secure cookie from the client is not matching with the server's identity");
+					throw new IllegalStateException("Invalid secure cookie passed.");
+				}
+
+				LOG.info("Secure cookie validation passed");
+
+				channelList.add(ctx.channel());
```

Comment:
It seems like we only have channels added, but not removed. Would this list increase indefinitely?